File: LeetCode/kth-largest-element-in-an-array-HEAP.cpp
// https://leetcode.com/problems/kth-largest-element-in-an-array/
// https://leetcode.com/problems/kth-largest-element-in-an-array/discuss/60309/C%2B%2B-STL-partition-and-heapsort

// MAX HEAP
class Solution
{
public:
    int findKthLargest(vector<int> &nums, int k)
    {
        priority_queue<int> pq(nums.begin(), nums.end());
        // here we basically removed the k - 1 larger elements
        while (k-- > 1)
        {
            pq.pop();
        }
        return pq.top();
    }
};

// MIN HEAP
class Solution
{
public:
    int findKthLargest(vector<int> &nums, int k)
    {
        priority_queue<int, vector<int>, greater<int>> pq;
        for (int i = 0; i < nums.size(); ++i)
        {
            pq.push(nums[i]);
            // here we only let the largest k elements of the array
            // remain in the heap, as we're removing all the smallest ones
            if (pq.size() > k)
            {
                pq.pop();
            }
        }
        // only k largest elements of the array remain
        // and the smallest of those will be the kth largest
        return pq.top();
    }
};

File: LeetCode/kth-largest-element-in-an-array-HEAP.java
// https://leetcode.com/problems/kth-largest-element-in-an-array/

class Solution {
  public int findKthLargest(int[] nums, int k) {
    // MIN HEAP: Smallest element first.
    PriorityQueue<Integer> heap = new PriorityQueue<>(nums.length, (n1, n2) -> n1 - n2);

    // Keep K largest elements in heap.
    // We maintain a heap of size K - and since this is a min heap
    // the order of elements will be from smallest to largest.
    // With the first element automatically being the K'th largest
    // element (largest element seen so far would be the last element).
    // Time complexity: O(N * log(K))
    for (int n : nums) {
      heap.add(n);
      if (heap.size() > k) {
        heap.poll();
      }
    }

    return heap.poll();
  }
}